new solver: prefer trivial builtin impls

lqd opened the pull request:
As discussed [on zulip](https://rust-lang.zulipchat.com/#narrow/channel/364551-t-types.2Ftrait-system-refactor/topic/needs_help.3A.20trivial.20builtin.20impls), this PR:
- adds a new `BuiltinImplSource::Trivial` source, and marks the `Sized` builtin impls as trivial
- prefers these trivial builtin impls in `merge_trait_candidates`

The comments can likely be wordsmithed a bit better, and I ~stole~ was inspired by the old solver ones. Let me know how you want them improved.

When enabling the new solver for tests, 3 UI tests now pass:
- `regions/issue-26448-1.rs` and its sibling `regions/issue-26448-2.rs` were rejected by the new solver but accepted by the old one
- and `issues/issue-42796.rs` where the old solver emitted some overflow errors in addition to the expected error

(For some reason one of these tests is run-pass, but I can take care of that another day)

r? lcnr

lqd committed:
new solver: prefer trivial builtin impls over where-clauses

for now, only builtin `Sized` impls are tracked as being `Trivial`

lcnr commented on compiler/rustc_type_ir/src/solve/mod.rs:
```diff
@@ -185,6 +185,9 @@ pub enum BuiltinImplSource {
     /// This can be removed when `feature(tuple_unsizing)` is stabilized, since we only
     /// use it to detect when unsizing tuples in hir typeck.
     TupleUnsizing,
+    /// A built-in impl that is considered trivial, without any nested requirements. They
+    /// are preferred over where-clauses, and we want to track them explicitly.
+    Trivial,
```
vibe: pls move them to the top of this enum

lqd replied:
👍

lcnr commented:
please add the following to at least one affected test
```rust
//@ revisions: current next
//@[next] compile-flags: -Znext-solver
//@ ignore-compare-mode-next-solver (explicit revisions)
```

r=me after nits

lcnr commented on compiler/rustc_next_trait_solver/src/solve/trait_goals.rs:
```diff
@@ -1194,7 +1195,30 @@ where
 
-        // FIXME: prefer trivial builtin impls
+        // We prefer trivial builtin candidates, i.e. builtin impls without any
+        // nested requirements, over all others. This is a fix for #53123 and
+        // prevents winnowing from accidentally extending the lifetime of a
+        // variable.
```
```suggestion
        // nested requirements, over all others. This is a fix for #53123 and
        // prevents where-bounds from accidentally extending the lifetime of a
        // variable.
```

lqd replied:
👍

lqd commented:
I added the explicit revisions to the two tests that didn't need blessing for the new solver (and fixed the run-pass I mentioned above), and were initially affected when inhibiting this very preference in the old solver.

@bors r=lcnr rollup

bors commented:
:pushpin: Commit 00844be421ff67872681cf9c4d1a4cb8d4e90891 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr 00844be421ff67872681cf9c4d1a4cb8d4e90891 -->
<!-- homu: {"type":"Approved","sha":"00844be421ff67872681cf9c4d1a4cb8d4e90891","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
